### PR TITLE
Optimize AOT precompilation in tpu-inference

### PR DESCRIPTION
**Pull request by igorts-git**
# Description

This PR optimizes Ahead-Of-Time (AOT) precompilation and dummy weight loading in the TPU inference runner to reduce HBM consumption and eliminate redundant compilation overhead.

### Key Changes

**Memory-efficient AOT Precompilation**: Replaces concrete dummy array allocations (`_create_dummy_tensor`) with lightweight `jax.ShapeDtypeStruct` during JIT function lowering (`fn.lower`). This avoids allocating High Bandwidth Memory (HBM) on the TPU for dummy inputs during compilation. As a result, the `MAX_PRECOMPILE_PROMPT_TOKENS` limit (previously capped at 1024) is removed, allowing precompilation of larger prompt token shapes without HBM exhaustion.

### Implementation Details
- Added `compile_only` to `_run_compilation` and `_precompile` methods in `CompilationManager` to support lowering and compiling JIT functions without adding them to warmup execution passes.
- Migrated placeholder tensor creations in `compilation_manager.py` to `jax.ShapeDtypeStruct`.

FIXES: b/528378191

# Tests

Manual invocations of MaxText `train_rl.py`.

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/runner/compilation_manager.py [hunk 1]
```diff
@@ -858,6 +868,7 @@ def _precompile_compute_pooling(self) -> None:
     def _precompile_sampling(self) -> None:
         logger.info("Compiling sampling with different input shapes.")
         hsize = self.runner.vocab_size
+        replicated_sharding = NamedSharding(self.runner.mesh, PartitionSpec())
```
Comment (wenxindongwork) [inline tpu_inference/runner/compilation_manager.py:933]: is this `replicated_sharding` necessary?

Comment (igorts-git) [reply tpu_inference/runner/compilation_manager.py:933]: It is used in line 905. My JetSki tells me that if we don't specify it we will get compilation cache mismatches.

**Review by wenxindongwork — APPROVED** [review]

**Review by jrplatin — APPROVED** [review]
